src/__tests__/unit/controllers/hooks/router-store/test.tsx

```diff
@@ -0,0 +1,306 @@
+import React from 'react';
+
+import { mount } from 'enzyme';
+import * as historyHelper from 'history';
+import { defaultRegistry } from 'react-sweet-state';
+import { act } from 'react-dom/test-utils';
+
+import { Router } from '../../../../../controllers';
+import {
+  useQueryParam,
+  usePathParam,
+} from '../../../../../controllers/hooks/router-store';
+import { getRouterStore } from '../../../../../controllers/router-store';
+
+const mockLocation = {
+  pathname: '/projects/123/board/456',
+  search: '?foo=hello&bar=world',
+  hash: '#hash',
+};
+
+const mockRoutes = [
+  {
+    path: '/projects/:projectId/board/:boardId',
+    component: () => <div>path</div>,
+    name: '',
+  },
+  {
+    path: '/blah',
+    component: () => <div>path</div>,
+    name: '',
+  },
+];
+
+const historyBuildOptions = {
+  initialEntries: [
+    `${mockLocation.pathname}${mockLocation.search}${mockLocation.hash}`,
+  ],
+};
+
+let history = historyHelper.createMemoryHistory(historyBuildOptions);
+let historyPushSpy = jest.spyOn(history, 'push');
+const nextTick = () => new Promise(resolve => setTimeout(resolve));
+
+const MockComponent = ({ children, ...rest }: any) => {
+  return children(rest);
+};
+
+describe('useQueryParam', () => {
+  beforeEach(() => {
+    history = historyHelper.createMemoryHistory(historyBuildOptions);
+    historyPushSpy = jest.spyOn(history, 'push');
+  });
+
+  afterEach(() => {
+    defaultRegistry.stores.clear();
+    jest.resetAllMocks();
+    jest.restoreAllMocks();
+  });
+
+  it('should return the right param value', () => {
+    mount(
+      <Router routes={mockRoutes} history={history}>
+        <MockComponent>
+          {() => {
+            const [param] = useQueryParam('foo');
+            expect(param).toEqual('hello');
+
+            return <div>I am a subscriber</div>;
+          }}
+        </MockComponent>
+      </Router>
+    );
+  });
+
+  it('should return undefined for non-existent params', () => {
```

Comment:
can you add a test for adding a new param? So it should return `undefined` in the first place, and then after calling `setParam` it should update the URL

Reply:
Good call. I've added a test to cover this case. Surprisingly it worked without any changes. Goes to show that we got the API design right.